### PR TITLE
q-gram distance

### PR DESCRIPTION
**Pull request by snim2**
Calculate a q-gram distance between strings. Needed by the Zhang-Shasha matcher.

Implementation based on the simmetrics library: https://github.com/Simmetrics/simmetrics

### CODE REVIEW

**Comment by snim2** [comment]
Ah, hang about, there are some clippy suggestions to go in. Sorry!

**Comment by snim2** [comment]
OK, will need squashing

**Comment by ltratt** [comment]
Do you want to squash this before I review?

**Comment by snim2** [comment]
Squashed!

File: src/lib/qgram.rs [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            return token_set;
+        } else if padded.len() < self.q as usize {
+            token_set.insert(padded);
+            return token_set;
+        }
+        // To parse a string into a q-gram set of tokens we move a `self.q`-wide
+        // sliding windows across the string.
+        let chars = padded.chars().collect::<Vec<char>>();
+        let last_qgram_index = chars.len() - self.q as usize;
+        for qgram_start in 0..=last_qgram_index {
```
Comment (ltratt) [inline src/lib/qgram.rs]: Stray `=`? Or maybe it does something I don't know about?

**Comment by ltratt** [comment]
The whole `#` thing seems weird to me -- is there a justification for this in the literature? I'd be curious to have a cut-down explanation!

File: src/lib/qgram.rs (hunk 1 above)
Comment (snim2) [reply src/lib/qgram.rs]: This is equivalent to `for (qgram_start = 0; i <= last_qgram_index; i++) { ... }` so the `..=` is like the Java `<=`.

**Comment by snim2** [comment]
So, when we perform a q-gram distance computation, we first tokenise the input string. If the string is `ABCDEF` then (without padding) the 3-gram tokenisation would be `ABC, BCD, CDE, DEF`. Notice that `C` appears three times in that list, but not so `A` or `F`. With padding, we would tokenise to the list `##A, #AB, ABC, BCD, CDE, DEF, EF#, F##` and every letter appears 3 times. Sometimes people use `^` and `$` for the start and end padding tokens, but we should probably use a token that we believe is less likely to appear in string literals and comments. I'm not sure that `#` is the best choice, but it was the default padding string in the simmetrics library that GT uses.

File: src/lib/qgram.rs (hunk 1 above)
Comment (ltratt) [reply src/lib/qgram.rs]: I didn't know that. I don't like it either (Rust, why so much syntax?!) but fair enough.

**Comment by ltratt** [comment]
What happens then if the user string *does* contain `#`? Does it throw the algorithm off a little or ... ?

File: src/lib/qgram.rs (hunk 1 above)
Comment (vext01) [reply src/lib/qgram.rs]: Gross!

**Comment by ltratt** [comment]
"Same result" being a boolean or ... ?

**Comment by snim2** [comment]
No, the distance is a float

**Comment by ltratt** [comment]
Hmm.

**Comment by ltratt** [comment]
Do we only ever compare pairs of strings at a time? If so, we could use a "fresh" character: i.e. a character `c` which is not in either of the two strings. With unicode this should be pretty easy to do.

**Comment by snim2** [comment]
Indeed we do only compare pairs of strings. I'm not sure what you mean about unicode though?

**Comment by ltratt** [comment]
Well, in theory, someone could give us two strings that between them use every possible unicode character. But those would be two long strings (4GiB between them), so I think we can assume that won't happen. What that means is that for two strings S1 and S2, it should be easy to find a character C that doesn't appear in either.

**Comment by ltratt** [comment]
Ping @snim2.

**Comment by snim2** [comment]
OK, this is ready for review. These last commits:

  * Default to a padding character that is highly unlikely to appear in real strings (the "replacement" char which is used to indicate a unicode decoding error)
  * If the default padding character does appear in a real string, we look for an alternative.

File: src/lib/qgram.rs [hunk 2]
```diff
@@ -83,15 +83,44 @@ impl QGram {
         distance
     }
 
+    /// Choose a padding character that does not appear in `input`.
+    ///
+    /// Use `self.padding` if possible.
+    /// Panics if it cannot find a suitable padding character.
+    fn choose_padding(&self, input: &str) -> String {
+        // Note that `first()` is not defined for the `Chars` type.
```
Comment (ltratt) [inline src/lib/qgram.rs]: Comment is unneeded.

File: src/lib/qgram.rs [hunk 3]
```diff
@@ -83,15 +83,44 @@ impl QGram {
...
+        if !input.contains(self.padding.chars().nth(0).unwrap()) {
```
Comment (ltratt) [inline src/lib/qgram.rs]: This makes this function O(2n) when we can make this O(n) by folding this into the `for` loop below and having a flag `found_padding_char` which is set to true if `c == padding_char`. [Hopefully that makes sense.]

**Comment by snim2** [comment]
OK, hopefully that addresses the comments.

**Comment by ltratt** [comment]
OK, I think we're ready to squash and rebase against master.

**Comment by ltratt** [comment]
@snim2 Is this ready?

**Comment by snim2** [comment]
Sorry, for got to say -- ready and waiting!

**Pull request merged**